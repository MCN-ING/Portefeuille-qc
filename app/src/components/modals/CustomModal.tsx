import { Button, ButtonType, testIdWithKey, useTheme } from '@hyperledger/aries-bifold-core'
import { useTranslation } from 'react-i18next'
import {
  Modal,
  StyleSheet,
  Text,
  TouchableOpacity,
  TouchableWithoutFeedback,
  useWindowDimensions,
  View,
} from 'react-native'
import { ScrollView } from 'react-native-gesture-handler'
import Icon from 'react-native-vector-icons/MaterialIcons'

import { hitSlop } from '../../constants'
import HeaderText from '../HeaderText'

export interface CustomModalProps {
  title: string
  onDismissPressed: () => void
  description?: string
  primary?: {
    label: string
    action: () => void
  }
  secondary?: {
    label: string
    action: () => void
  }
}

export const CustomModal = ({ title, description, primary, secondary, onDismissPressed }: CustomModalProps) => {
  const { height, width } = useWindowDimensions()
  const { TextTheme, ColorPallet } = useTheme()
  const { t } = useTranslation()

  const styles = StyleSheet.create({
    modalCenter: {
      flex: 1,
      justifyContent: 'center',
      alignItems: 'center',
      paddingHorizontal: 20,
      paddingVertical: 100,
      backgroundColor: ColorPallet.notification.popupOverlay,
      minHeight: height,
      maxHeight: height,
      minWidth: width,
    },
    container: {
      padding: 20,
      minWidth: width - 2 * 25,
      backgroundColor: ColorPallet.brand.primaryBackground,
      shadowColor: ColorPallet.grayscale.darkGrey,
      shadowOffset: {
        width: 3,
        height: 3,
      },
      shadowOpacity: 0.6,
      elevation: 5,
    },
    scrollViewStyle: {
      flexGrow: 0,
    },
    titleContainer: {
      flexDirection: 'row',
      justifyContent: 'space-between',
    },
    bodyText: {
      ...TextTheme.normal,
      paddingVertical: 16,
      color: ColorPallet.notification.infoText,
    },
    ActionContainer: {
      paddingTop: 10,
    },
  })

  return (
    <Modal transparent accessibilityViewIsModal>
      <TouchableOpacity onPress={onDismissPressed} accessible={false}>
        <View style={styles.modalCenter}>
          <TouchableWithoutFeedback accessible={false}>
            <View style={styles.container}>
              <View style={styles.titleContainer}>
                <HeaderText title={title} />
                <TouchableOpacity
                  onPress={onDismissPressed}
                  testID={testIdWithKey('Close')}
                  accessibilityLabel={t('Global.Close')}
                  accessibilityRole={'button'}
                  hitSlop={hitSlop}
                >
                  <Icon name={'clear'} size={30} color={ColorPallet.notification.infoIcon} />
                </TouchableOpacity>
              </View>
              <ScrollView showsVerticalScrollIndicator={false} style={styles.scrollViewStyle}>
                <View onStartShouldSetResponder={() => true}>
                  <Text style={styles.bodyText} testID={testIdWithKey('BodyText')}>
                    {description}
                  </Text>
                </View>
              </ScrollView>
              <View style={styles.ActionContainer}>
                <Button
                  title={primary?.label ?? secondary?.label ?? t('Global.Close')}
                  accessibilityLabel={primary?.label ?? secondary?.label ?? t('Global.Close')}
                  testID={testIdWithKey('ModalPrimaryAction')}
                  buttonType={ButtonType.Primary}
                  onPress={primary?.action ?? secondary?.action ?? onDismissPressed}
                />
              </View>
              {primary && secondary && (
                <View style={styles.ActionContainer}>
                  <Button
                    title={secondary.label}
                    accessibilityLabel={secondary.label}
                    testID={testIdWithKey('ModalSecondaryAction')}
                    onPress={secondary.action}
                    buttonType={ButtonType.Secondary}
                  />
                </View>
              )}
            </View>
          </TouchableWithoutFeedback>
        </View>
      </TouchableOpacity>
    </Modal>
  )
}
