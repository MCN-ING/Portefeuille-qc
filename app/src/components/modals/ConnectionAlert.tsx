import { Link, useTheme, Stacks as BifoldStacks, Screens as BifoldScreens } from '@hyperledger/aries-bifold-core'
import UnorderedList from '@hyperledger/aries-bifold-core/App/components/misc/UnorderedList'
import { useNavigation } from '@react-navigation/native'
import { StackNavigationProp } from '@react-navigation/stack'
import React, { useState } from 'react'
import { useTranslation } from 'react-i18next'
import { StyleSheet, Text, TouchableOpacity, View } from 'react-native'
import Icon from 'react-native-vector-icons/MaterialCommunityIcons'

import { hitSlop } from '../../constants'
import { RootStackParams } from '../../navigators/navigators'

import { CustomModal } from './CustomModal'

interface ConnectionAlertProps {
  connectionID?: string
}

const ConnectionAlert: React.FC<ConnectionAlertProps> = ({ connectionID }) => {
  const { t } = useTranslation()
  const { ColorPallet, TextTheme } = useTheme()
  const [infoCardVisible, setInfoCardVisible] = useState(false)

  const settingsNavigation = useNavigation<StackNavigationProp<RootStackParams>>()

  const styles = StyleSheet.create({
    notifyTextContainer: {
      flex: 1,
      paddingVertical: 15,
      marginVertical: 15,
    },
    row: {
      flexDirection: 'row',
    },
    notifyTitle: {
      ...TextTheme.title,
      marginBottom: 5,
    },
    notifyText: {
      ...TextTheme.normal,
      marginVertical: 5,
    },
    informationIcon: {
      color: ColorPallet.notification.infoIcon,
      marginLeft: 10,
    },
  })

  const toggleInfoCard = () => setInfoCardVisible(!infoCardVisible)

  const navigateToContacts = () => {
    toggleInfoCard()
    settingsNavigation.navigate(BifoldStacks.ContactStack, { screen: BifoldScreens.Contacts })
  }

  return (
    <View style={styles.notifyTextContainer}>
      <View style={styles.row}>
        <Text style={styles.notifyTitle}>{t('ConnectionAlert.AddedContacts')}</Text>
        <TouchableOpacity
          testID={t('Global.Info')}
          accessibilityLabel={t('ConnectionAlert.AddedContacts')}
          accessibilityRole={'button'}
          onPress={toggleInfoCard}
          hitSlop={hitSlop}
        >
          <Icon name={'information'} size={30} style={styles.informationIcon} />
        </TouchableOpacity>
      </View>
      {infoCardVisible && (
        <CustomModal
          title={t('ConnectionAlert.AddedContacts')}
          description={
            <View style={styles.notifyTextContainer}>
              <Text style={[styles.notifyText, { marginBottom: 16 }]}>{t('ConnectionAlert.PopupIntro')}</Text>
              <UnorderedList
                unorderedListItems={[t('ConnectionAlert.PopupPoint1'), t('ConnectionAlert.PopupPoint2')]}
              />
              <Text style={[styles.notifyText, { marginTop: 16 }]}>
                {t('ConnectionAlert.SettingsInstruction1')}
                <Link onPress={navigateToContacts} linkText={t('ConnectionAlert.ContactsLink')} />
                {t('ConnectionAlert.SettingsInstruction2')}
              </Text>
            </View>
          }
          primary={{
            label: t('ConnectionAlert.PopupExit'),
            action: toggleInfoCard,
          }}
          onDismissPressed={toggleInfoCard}
        />
      )}
      <Text style={styles.notifyText}>
        {connectionID
          ? t('ConnectionAlert.NotificationBodyWithIssuer', { issuerName: connectionID })
          : t('ConnectionAlert.NotificationBodyWithoutIssuer')}
      </Text>
    </View>
  )
}

export default ConnectionAlert
