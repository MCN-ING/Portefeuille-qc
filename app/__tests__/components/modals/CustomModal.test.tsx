import { testIdWithKey } from '@hyperledger/aries-bifold-core'
import { render, fireEvent } from '@testing-library/react-native'
import React from 'react'
import { CustomModal } from '../../../src/components/modals/CustomModal'

const title = 'Test Title'
const description = 'Lorem ipsum sit dolar amet'

describe('DismissibleCustomModal Component', () => {
  test('Renders correctly without call to action', () => {
    const onDismissPressed = jest.fn()
    const tree = render(<CustomModal title={title} description={description} onDismissPressed={onDismissPressed} />)

    expect(tree).toMatchSnapshot()
  })

  test('Renders correctly with primary call to action', () => {
    const onDismissPressed = jest.fn()
    const onCallToActionPressed = jest.fn()
    const tree = render(
      <CustomModal
        title={title}
        description={description}
        onDismissPressed={onDismissPressed}
        primary={{ label: 'Okay', action: onCallToActionPressed }}
      />
    )

    expect(tree).toMatchSnapshot()
  })

  test('Triggers call to action on primary press', () => {
    const onDismissPressed = jest.fn()
    const onCallToActionPressed = jest.fn()
    const { getByText } = render(
      <CustomModal
        title={title}
        description={description}
        onDismissPressed={onDismissPressed}
        primary={{ label: 'Okay', action: onCallToActionPressed }}
      />
    )
    const okayButton = getByText('Okay')
    fireEvent(okayButton, 'press')

    expect(onCallToActionPressed).toBeCalledTimes(1)
  })

  test('Triggers call to action on secondary press', () => {
    const onDismissPressed = jest.fn()
    const onCallToActionPressed = jest.fn()
    const { getByText } = render(
      <CustomModal
        title={title}
        description={description}
        onDismissPressed={onDismissPressed}
        primary={{ label: 'Okay', action: () => {} }}
        secondary={{ label: 'Cancel', action: onCallToActionPressed }}
      />
    )
    const okayButton = getByText('Cancel')
    fireEvent(okayButton, 'press')

    expect(onCallToActionPressed).toBeCalledTimes(1)
  })

  test('Triggers dismiss on press', () => {
    const onDismissPressed = jest.fn()
    const onCallToActionPressed = jest.fn()
    const { getByTestId } = render(
      <CustomModal
        title={title}
        description={description}
        onDismissPressed={onDismissPressed}
        primary={{ label: 'Okay', action: onCallToActionPressed }}
      />
    )
    const closeButton = getByTestId(testIdWithKey('Close'))
    fireEvent(closeButton, 'press')

    expect(onDismissPressed).toBeCalledTimes(1)
  })
})
