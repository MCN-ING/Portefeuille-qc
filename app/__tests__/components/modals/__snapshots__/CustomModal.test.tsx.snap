// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DismissibleCustomModal Component Renders correctly with primary call to action 1`] = `
<Modal
  accessibilityViewIsModal={true}
  hardwareAccelerated={false}
  transparent={true}
  visible={true}
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={false}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "backgroundColor": "rgba(0, 0, 0, 0.5)",
          "flex": 1,
          "justifyContent": "center",
          "maxHeight": 1334,
          "minHeight": 1334,
          "minWidth": 750,
          "paddingHorizontal": 20,
          "paddingVertical": 100,
        }
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessible={false}
        focusable={false}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "backgroundColor": "#000000",
            "elevation": 5,
            "minWidth": 700,
            "padding": 20,
            "shadowColor": "#313132",
            "shadowOffset": {
              "height": 3,
              "width": 3,
            },
            "shadowOpacity": 0.6,
          }
        }
      >
        <View
          style={
            {
              "flexDirection": "row",
              "justifyContent": "space-between",
            }
          }
        >
          <View>
            <Text
              accessibilityRole="header"
              style={
                [
                  {
                    "color": "#FFFFFF",
                    "fontSize": 32,
                    "fontWeight": "bold",
                    "lineHeight": 32,
                  },
                ]
              }
            >
              Test Title
            </Text>
            <View
              style={
                {
                  "backgroundColor": "#FCBA19",
                  "height": 4,
                  "width": 48,
                }
              }
            />
          </View>
          <View
            accessibilityLabel="Global.Close"
            accessibilityRole="button"
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            hitSlop={
              {
                "bottom": 44,
                "left": 44,
                "right": 44,
                "top": 44,
              }
            }
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "opacity": 1,
              }
            }
            testID="com.ariesbifold:id/Close"
          >
            <Text
              allowFontScaling={false}
              selectable={false}
              style={
                [
                  {
                    "color": "#0099FF",
                    "fontSize": 30,
                  },
                  undefined,
                  {
                    "fontFamily": "Material Icons",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  {},
                ]
              }
            >
              î…Œ
            </Text>
          </View>
        </View>
        <RCTScrollView
          showsVerticalScrollIndicator={false}
          style={
            {
              "flexGrow": 0,
            }
          }
        >
          <View>
            <View
              onStartShouldSetResponder={[Function]}
            >
              <Text
                style={
                  {
                    "color": "#FFFFFF",
                    "fontSize": 18,
                    "fontWeight": "normal",
                    "paddingVertical": 16,
                  }
                }
              >
                Lorem ipsum sit dolar amet
              </Text>
            </View>
          </View>
        </RCTScrollView>
        <View
          style={
            {
              "paddingTop": 10,
            }
          }
        >
          <View
            accessibilityLabel="Okay"
            accessibilityRole="button"
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": false,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "backgroundColor": "#42803E",
                "borderRadius": 4,
                "opacity": 1,
                "padding": 16,
              }
            }
            testID="com.ariesbifold:id/ModalPrimaryAction"
          >
            <View
              style={
                {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "justifyContent": "center",
                }
              }
            >
              <Text
                style={
                  [
                    {
                      "color": "#FFFFFF",
                      "fontSize": 18,
                      "fontWeight": "bold",
                      "textAlign": "center",
                    },
                    false,
                    false,
                    false,
                  ]
                }
              >
                Okay
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</Modal>
`;

exports[`DismissibleCustomModal Component Renders correctly without call to action 1`] = `
<Modal
  accessibilityViewIsModal={true}
  hardwareAccelerated={false}
  transparent={true}
  visible={true}
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={false}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "backgroundColor": "rgba(0, 0, 0, 0.5)",
          "flex": 1,
          "justifyContent": "center",
          "maxHeight": 1334,
          "minHeight": 1334,
          "minWidth": 750,
          "paddingHorizontal": 20,
          "paddingVertical": 100,
        }
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessible={false}
        focusable={false}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "backgroundColor": "#000000",
            "elevation": 5,
            "minWidth": 700,
            "padding": 20,
            "shadowColor": "#313132",
            "shadowOffset": {
              "height": 3,
              "width": 3,
            },
            "shadowOpacity": 0.6,
          }
        }
      >
        <View
          style={
            {
              "flexDirection": "row",
              "justifyContent": "space-between",
            }
          }
        >
          <View>
            <Text
              accessibilityRole="header"
              style={
                [
                  {
                    "color": "#FFFFFF",
                    "fontSize": 32,
                    "fontWeight": "bold",
                    "lineHeight": 32,
                  },
                ]
              }
            >
              Test Title
            </Text>
            <View
              style={
                {
                  "backgroundColor": "#FCBA19",
                  "height": 4,
                  "width": 48,
                }
              }
            />
          </View>
          <View
            accessibilityLabel="Global.Close"
            accessibilityRole="button"
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            hitSlop={
              {
                "bottom": 44,
                "left": 44,
                "right": 44,
                "top": 44,
              }
            }
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "opacity": 1,
              }
            }
            testID="com.ariesbifold:id/Close"
          >
            <Text
              allowFontScaling={false}
              selectable={false}
              style={
                [
                  {
                    "color": "#0099FF",
                    "fontSize": 30,
                  },
                  undefined,
                  {
                    "fontFamily": "Material Icons",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  {},
                ]
              }
            >
              î…Œ
            </Text>
          </View>
        </View>
        <RCTScrollView
          showsVerticalScrollIndicator={false}
          style={
            {
              "flexGrow": 0,
            }
          }
        >
          <View>
            <View
              onStartShouldSetResponder={[Function]}
            >
              <Text
                style={
                  {
                    "color": "#FFFFFF",
                    "fontSize": 18,
                    "fontWeight": "normal",
                    "paddingVertical": 16,
                  }
                }
              >
                Lorem ipsum sit dolar amet
              </Text>
            </View>
          </View>
        </RCTScrollView>
        <View
          style={
            {
              "paddingTop": 10,
            }
          }
        >
          <View
            accessibilityLabel="Global.Close"
            accessibilityRole="button"
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": false,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "backgroundColor": "#42803E",
                "borderRadius": 4,
                "opacity": 1,
                "padding": 16,
              }
            }
            testID="com.ariesbifold:id/ModalPrimaryAction"
          >
            <View
              style={
                {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "justifyContent": "center",
                }
              }
            >
              <Text
                style={
                  [
                    {
                      "color": "#FFFFFF",
                      "fontSize": 18,
                      "fontWeight": "bold",
                      "textAlign": "center",
                    },
                    false,
                    false,
                    false,
                  ]
                }
              >
                Global.Close
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</Modal>
`;
